File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/oo/sherlock_and_squares_oo.cs
// Source: https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

using System;

public class Solution
{
    public static void Main()
    {
        int nTestCases = _ReadANumber();
        List<int> output = new List<int>(new int[nTestCases]);

        for (int i = 0; i < nTestCases; i++)
        {
            List<int> arrayInput = _ReadAnArray();
            int begin = arrayInput.First();
            int end = arrayInput.Last();

            SherlockAndSquares obj = new SherlockAndSquares(begin, end);
            output[i] = obj.GetQuantityOfSquareIntegers();
        }

        PrintArray(output);
    }

        private static int _ReadANumber()
        {
            int number = int.Parse(Console.ReadLine());
            return number;
        }

        private static List<int> _ReadAnArray()
        {
            List<int> array = Console.ReadLine().Split(" ").Select(int.Parse).ToList();
            return array;
        }

        public static void PrintArray(List<int> array)
        {
            array.ForEach(Console.WriteLine);
        }
}

    public class SherlockAndSquares
    {
        private int _Begin;
        private int _End;
        private int _QuantityOfSquareIntegers;

        public SherlockAndSquares(int begin, int end)
        {
            _Begin = begin;
            _End = end;
            _QuantityOfSquareIntegers = 0;

            _CalculateNumberOfSquareIntegersInRange();
        }

            private void _CalculateNumberOfSquareIntegersInRange()
            {
                for (int number = _FindFirstSquareIntegerFromBeginNumber(); _IsPowerOf2OfNumberLessThanOrEqualToEndNumber(number); number++)
                    _QuantityOfSquareIntegers++;
            }

                private int _FindFirstSquareIntegerFromBeginNumber()
                {
                    double squareRootNumber = Math.Sqrt(_Begin);
                    int squareInteger = _GetSquareIntegerFromNumber(squareRootNumber);
                    return squareInteger;
                }

                    private int _GetSquareIntegerFromNumber(double number)
                    {
                        int integerPart = (int) Math.Floor(number);
                        return number == integerPart ? integerPart : integerPart + 1;
                    }

                private bool _IsPowerOf2OfNumberLessThanOrEqualToEndNumber(int number)
                {
                    return Math.Pow(number, 2) <= _End;
                }

        public int GetQuantityOfSquareIntegers()
        {
            return _QuantityOfSquareIntegers;
        }
    }

File: hackerrank/problem solving/algorithms/2 - implementation/32 - sherlock and squares/oo/sherlock_and_squares_oo.java
// Source: https://www.hackerrank.com/challenges/sherlock-and-squares/problem?isFullScreen=true

import java.util.Scanner;
import java.util.List;
import java.util.Arrays;
import java.lang.Math;

public class Solution {
    private static Scanner _scanner;

    public static void main(String[] args) {
        _scanner = new Scanner(System.in);

        final int N_TEST_CASES = _readANumber();
        List<Integer> output = Arrays.asList(new Integer[N_TEST_CASES]);

        for (int i = 0; i < N_TEST_CASES; i++) {
            final int BEGIN = _readANumber();
            final int END = _readANumber();

            final SherlockAndSquares OBJ = new SherlockAndSquares(BEGIN, END);
            output.set(i, OBJ.getQuantityOfSquareIntegers());
        }

        printArray(output);
    }

        private static int _readANumber() {
            final int NUMBER = _scanner.nextInt();
            return NUMBER;
        }

        public static void printArray(List<Integer> array) {
            array.forEach(System.out::println);
        }
}

    class SherlockAndSquares {
        private int _begin;
        private int _end;
        private int _quantityOfSquareIntegers;

        public SherlockAndSquares(int begin, int end) {
            _begin = begin;
            _end = end;
            _quantityOfSquareIntegers = 0;

            _calculateNumberOfSquareIntegersInRange();
        }

            private void _calculateNumberOfSquareIntegersInRange() {
                for (int number = _findFirstSquareIntegerFromBeginNumber(); _isPowerOf2OfNumberLessThanOrEqualToEndNumber(number); number++)
                    _quantityOfSquareIntegers++;
            }

            private int _findFirstSquareIntegerFromBeginNumber() {
                final double SQUARE_ROOT_NUMBER = Math.sqrt(_begin);
                final int SQUARE_INTEGER = _getSquareIntegerFromNumber(SQUARE_ROOT_NUMBER);
                return SQUARE_INTEGER;
            }

                private int _getSquareIntegerFromNumber(double number) {
                    final int INTEGER_PART = (int) Math.floor(number);
                    return number == INTEGER_PART ? INTEGER_PART : INTEGER_PART + 1;
                }

            private boolean _isPowerOf2OfNumberLessThanOrEqualToEndNumber(int number) {
                return Math.pow(number, 2) <= _end;
            }

        public int getQuantityOfSquareIntegers() {
            return _quantityOfSquareIntegers;
        }
    }
